dll/win32/shell32/dialogs/filetypes.cpp

```diff
@@ -56,204 +158,256 @@ DeleteExt(HWND hwndDlg, LPCWSTR pszExt)
 
-    // Open the "DefaultIcon" registry key
     HKEY hDefIconKey;
     LONG nResult = RegOpenKeyExW(hKey, L"DefaultIcon", 0, KEY_READ, &hDefIconKey);
     if (nResult != ERROR_SUCCESS)
         return FALSE;
 
     // Get the icon location
-    WCHAR szLocation[MAX_PATH + 32] = { 0 };
+    WCHAR szLocation[ICONLOCATION_CCHMAX];
     DWORD dwSize = sizeof(szLocation);
     nResult = RegQueryValueExW(hDefIconKey, NULL, NULL, NULL, LPBYTE(szLocation), &dwSize);
+    szLocation[_countof(szLocation) - 1] = UNICODE_NULL;
 
     RegCloseKey(hDefIconKey);
-
-    if (nResult != ERROR_SUCCESS || szLocation[0] == 0)
+    if (nResult != ERROR_SUCCESS || !szLocation[0])
         return FALSE;
 
-    return GetFileTypeIconsEx(Entry, szLocation);
+    return GetFileTypeIconsEx(Entry, szLocation, IconSize);
+}
+
+static LPCWSTR
+GetProgramPath(PFILE_TYPE_ENTRY Entry)
+{
+    if (!Entry->ProgramPath[1] && !Entry->ProgramPath[0])
+    {
+        DWORD cch = _countof(Entry->ProgramPath);
+        if (FAILED(AssocQueryStringW(ASSOCF_INIT_IGNOREUNKNOWN, ASSOCSTR_EXECUTABLE,
+                                     Entry->FileExtension, NULL, Entry->ProgramPath, &cch)))
+        {
+            Entry->ProgramPath[0] = UNICODE_NULL;
+            Entry->ProgramPath[1] = TRUE;
+        }
+    }
+    return Entry->ProgramPath;
 }
 
 static BOOL
 QueryFileDescription(LPCWSTR ProgramPath, LPWSTR pszName, INT cchName)
 {
-    SHFILEINFOW FileInfo = { 0 };
-    if (SHGetFileInfoW(ProgramPath, 0, &FileInfo, sizeof(FileInfo), SHGFI_DISPLAYNAME))
+    SHFILEINFOW fi;
+    fi.szDisplayName[0] = UNICODE_NULL;
+    if (SHGetFileInfoW(ProgramPath, 0, &fi, sizeof(fi), SHGFI_DISPLAYNAME))
     {
-        StringCchCopyW(pszName, cchName, FileInfo.szDisplayName);
+        StringCchCopyW(pszName, cchName, fi.szDisplayName);
         return TRUE;
     }
-
     return !!GetFileTitleW(ProgramPath, pszName, cchName);
 }
 
+static LPCWSTR
+GetAppName(PFILE_TYPE_ENTRY Entry)
+{
+    if (!Entry->AppName[1] && !Entry->AppName[0])
+    {
+        LPCWSTR exe = GetProgramPath(Entry);
+        if (!*exe || !QueryFileDescription(exe, Entry->AppName, _countof(Entry->AppName)))
+        {
+            Entry->AppName[0] = UNICODE_NULL;
+            Entry->AppName[1] = TRUE;
+        }
+    }
+    return Entry->AppName;
+}
+
+static LPWSTR
+GetTypeName(PFILE_TYPE_ENTRY Entry, PFILE_TYPE_GLOBALS pG)
+{
+    if (!Entry->FileDescription[1] && !Entry->FileDescription[0])
+    {
+        Entry->FileDescription[1] = TRUE;
+        if (Entry->IsExtension())
+        {
+            SHFILEINFOW fi;
+            if (SHGetFileInfoW(Entry->FileExtension, 0, &fi, sizeof(fi), SHGFI_TYPENAME |
+                               SHGFI_USEFILEATTRIBUTES) && *fi.szTypeName)
+            {
+                StringCchCopyW(Entry->FileDescription, _countof(Entry->FileDescription), fi.szTypeName);
+            }
+            else
+            {
+                // FIXME: Remove this hack when SHGetFileInfo is fixed
```

Comment:
Do we know what needs to be fixed in that function?

Reply:
Extensions without a progid (.asm for example) are failing SHGetFileInfo. I'll update the comment so we remember what the problem is.